Ena 3490

haseeb-gh opened the pull request.

haseeb-gh committed:
ENA-3490

- Updated the code to work with latest htsjdk version APIs that were modified.
- Code that used those APIs that were entirely removed in the new htsjdk version is still broken.

haseeb-gh committed:
ENA-3490

- refactored fastq converter to use samreader instead of low level api.
- old code is still present but commented out.

haseeb-gh committed:
ENA-3490

- added the skipping of non primary alignments and non pf reads.
- refactored code.

haseeb-gh committed:
ENA-3490

- created a small test utility class for comparing fastq file data from old and new version.

haseeb-gh committed:
ENA-3490

- refactored utility tests.
- added comments.

haseeb-gh committed:
ENA-3490

- added more tests.

haseeb-gh committed:
ENA-3490

- removed tests. created a new utility application instead.
- added todo markers.

haseeb-gh committed:
ENA-3490

- removed commented out/testing/unneeded code.
- added comments.
- updated version number.

haseeb-gh committed:
ENA-3490

- removed more unnecesary code.

raskoleinonen commented on src/main/java/htsjdk/samtools/cram/encoding/reader/ReaderToFastq.java:
```diff
@@ -34,26 +34,26 @@ public ReaderToFastq(int bufSize) {
 	}
 
 	@Override
-	protected void writeRead(byte[] name, int flags, byte[] bases, byte[] scores) {
```
Let's remove the ReaderToFastq class. We checked that it is not used anywhere: https://github.com/search?q=org%3Aenasequence+ReaderToFastq&type=Code

raskoleinonen commented on src/main/java/htsjdk/samtools/cram/encoding/reader/AbstractFastqReader.java:
```diff
@@ -15,43 +15,12 @@
-
-public abstract class AbstractFastqReader extends AbstractReader {
-	private ReadFeatureBuffer rfBuf = new ReadFeatureBuffer();
+public abstract class AbstractFastqReader{
```
As we remove the ReaderToFastq class, let's remove AbstractFastqReader and merge its functionality into MultiFastqOutputter.

raskoleinonen commented on src/main/java/htsjdk/samtools/cram/encoding/reader/MultiFastqOutputter.java:
```diff
@@ -20,14 +27,11 @@
+ * Makes use of a collection based caching implementation that may produce results in which a paired read is incorrectly written
+ * to the unpaired file if the two mates are located far away from each other in the collection.
+ */
 public class MultiFastqOutputter extends AbstractFastqReader {
```
As we remove the ReaderToFastq class, let's remove AbstractFastqReader and merge its functionality into MultiFastqOutputter. We should also change the package from htsjdk.samtools.cram.encoding.reader to e.g. src/main/java/uk/ac/ebi/ena/readtools/fastq.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/cram/Cram2Fastq.java:
```diff
@@ -15,48 +15,46 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.zip.GZIPOutputStream;
+
 public class Cram2Fastq {
```
I would rename this package to sam and the class to Sam2Fastq as the class is intended to work on SAM, BAM, and CRAM files. We should have a unit tests for BAM and CRAM files that generate 1, 2 and 3 Fastqs. Small files are ok for these tests.

For direct integration with sra-stages instead of main2() we could have Sam2Fastq constructor that takes the params and then call create() on the instance. Or just rename main2() to create( Params params ) or something similar. In either case we could leave the Params and the main() as they are now to support existing command line use.

haseeb-gh replied:
The constructor/Param change has been made. I have also added new tests but finding a small enough cram file that does not require downloading huge data from the cram reference registry is going to take a little time.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/cram/Cram2Bam.java:
```diff
@@ -79,7 +79,7 @@ public static void main(String[] args) throws IOException, IllegalArgumentExcept
 			log.warn("No reference file specified, remote access over internet may be used to download public sequences. ");
 
-		InputStream is = null;
+		SeekableFileStream is = null;
```
Cram2Bam functionality should not be needed and it is best to remove the Cram2Bam class.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/cram/Cram2Fastq.java:
```diff
@@ -274,14 +264,9 @@ public CollatingDumper(InputStream cramIS, CRAMReferenceSource referenceSource,
 
 		@Override
 		protected AbstractFastqReader newReader() {
```
After we remove AbstractFastqReader then the constructor method should be renamed to createMultiFastqOutputter() or similar. Looks very confusing to the average person as we are supposedly creating a 'reader' but are actually creating a fastq 'writer'.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/cram/Cram2Fastq.java:
```diff
@@ -183,57 +181,49 @@ public Dumper(InputStream cramIS, CRAMReferenceSource referenceSource, int nofSt
-			cramHeader = CramIO.readCramHeader(cramIS);			
+			cramHeader = CramIO.readCramHeader(cramIS);
 
 			reader = newReader();
```
Confusing use of variable name 'reader'. Best just to declare here:
MultiFastqOutputter fastqOutputter = createMultiFastqOutputter();

Note that newReader() has been renamed to createMultiFastqOutputter()

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/utils/CramDumpVersionResultsValidator.java:
```diff
@@ -0,0 +1,257 @@
+import java.util.TreeMap;
+import java.util.function.Consumer;
+
+public class CramDumpVersionResultsValidator {
```
A more descriptive name for the class might be Sam2FastqStatsComparator (I am assuming here that we rename Cram2Fastq as Sam2Fastq). We are comparing stats between two sets of Fastqs generated by (equivalent of) Sam2Fastq.

This class might actually benefit from JCommander with description of the purpose of the class and its input parameters. However, this is definitely a low priority.

raskoleinonen commented on src/test/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/refs/CramReferenceInfoTest.java:
```diff
@@ -7,13 +7,15 @@
 public class 
 CramReferenceInfoTest 
 {
+    @Ignore("Fails with an invalid sequence name exception. Sequence name validation rules have been updated in updated versions of htsjdk.")
```
Why does the test fail? Is it because the CRAM is unaligned or because there is something wrong with the read names in the CRAM? Can we use another CRAM file to make the test work?

haseeb-gh replied:
the error that gets thrown is:

Sequence name * doesn't match regex: '[0-9A-Za-z!#$%&+./:;?@^_|~-][0-9A-Za-z!#$%&*+./:;=?@^_|~-]*'

Newer version of htsjdk added this sequence name regex that was not present before. The cram file that is being tested here contains a (or more) sequence that has an invalid name. Good idea about using a different file for the test. You think can point me to small enough unaligned cram file?

raskoleinonen replied:
Is it possible to ignore this sequence name check in htsjdk? We don't index information about CRAMs being unaligned/aligned so we would need to loop through small CRAM files to find files that are unaligned. It would be very useful actually to have at least one one such file around that we can read with htsjdk as it might come handy in other tests as well later.

haseeb-gh replied:
Unfortunately it is not possible to skip/turn off this valdiation check. I tried looking in htsjdk's source code as well as searched the internet but found nothing. I also tried to see if it might be quicker to just edit the cram file and change the sequence name to something else but the tools I know also use htsjdk underneath and they also threw the same error when I attempted to use them. There might be another way to avoid this by using Java Reflection API to change the regex pattern that htsjdk uses but it might require a little RnD on my part.

haseeb-gh replied:
Managed to use Java Reflection API to set a dummy matcher for the test.

raskoleinonen commented on src/test/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/refs/CramReferenceInfoTest.java:
```diff
@@ -7,13 +7,15 @@
 import uk.ac.ebi.ena.readtools.webin.cli.rawreads.BamScannerTest;
 
 public class 
 CramReferenceInfoTest 
```
Not sure if this class is in the correct package.

haseeb-gh replied:
There are a bunch of classes inside the package uk.ac.ebi.ena.readtools.webin.cli.rawreads. These were all here from before. Any suggestions on where should we move them or whether it ll be better to rename the package to something else ?

raskoleinonen replied:
There are two clients for the readtools project. webin-cli and sra-stages. The uk.ac.ebi.ena.readtools.webin.cli.rawreads package name suggests that it is used by webin-cli. Please check which classes are used by webin-cli and sra-stages before renaming/moving packages/classes so that we don't break these build unnecessarily. The package name uk.ac.ebi.ena.readtools.webin.cli.rawreads would be ok for classes that are indeed used (I hope only) by webin-cli. The readtools is intended to be an 'API' project used by webin-cli and sra-stages and as with any 'API' project we need to understand what is the public API that needs to be stable and what is the other stuff that can change without risking breaking the clients. E.g. Sam2Fastq will be part of the public API and once it is used directly by Eugene it should remain stable unless API changes are discussed and agreed with the client(s) .

haseeb-gh replied:
So I checked and found that none of the two repositories use any class from this package. I made sure of it by building them two locally with the new version of readtools. I think it might be ok to just remove the package and all of the class in there completely.

haseeb-gh replied:
although the classes in this packages are not being used by either of the other two repositories. they are however, being used by another class uk.ac.ebi.ena.readtools.validator.ReadsValidator inside this repo that is being used by webin-cli. so, as agreed by the team, we are going to keep the package as it is for now.

raskoleinonen reviewed (commented):
- I would check that the packages are appropriately named and not e.g. in htsjkd namespace.
- We should focus on the test coverage in the next review.
- We need to add apache 2 license headers. Check e.g. webin-cli for a Gradle plugin how to do this.
- Once finished with reviews we should run the code through IntelliJ cleanup (remove unused throws, imports etc. [saw some of these today]) and change the coding style for any classes we have changed to use Google Java coding style (there is a gradle plugin or this as well).

haseeb-gh committed:
ENA-3490

- code review followup.
- some changes still incomplete.

haseeb-gh committed:
ENA-3490

- updated fastq score generation logic.

haseeb-gh committed:
ENA-3490

- updated sam2fastq tests.
- added spotless plugin for putting license text in java files.

haseeb-gh commented:
@raskoleinonen please review all of the suggested changes that have been added to the PR. the gitlab ci check is failing because of the unaligned cram test. it seems like the Java Reflection API features I have added to that test are not taking effect. this could be related the JVM configuration that is spun by gitlab ci to run the build. i would say its probably ok to ignore this for as i have made sure that all the tests are passing on my local setup. but if it is bothering then let me know and i can see what we can do about it.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/RawReadsFile.java:
```diff
@@ -1,284 +1,284 @@
+    {
+        fastq( "fastq" ),
+        bam( "bam"),
+        cram("cram2fastq");
```
Could this be an error? Should the string literal be 'cram' instead of 'cram2fastq'?

haseeb-gh committed:
ENA-3490

- fixed an overlooked potential issue.
- disabled the unaligned cram file test.
